lib/puppet/provider/mount/parsed.rb

```diff
@@ -272,6 +272,49 @@ def self.mountinstances
     instances
   end
 
+  # Sort fstab entries so that
+  # a mount point is located inside another mount point appears later and
+  # a device that is located inside a mount point appears later.
+  #
+  # "Stable opportunistic insertion sort"
+  # Copy unsorted list A to sorted B by repeatedly
+  # * removing the first element from A
+  # * compare with each element in B, from first to last
+  # * insert before current element in B if smaller
+  # * insert at the end of B otherwise
+  # * finish if A is empty
+  # * repeat
+  #
+  # Stabilizes itself by inserting as late as possible.
+  #
+  # A record is "smaller" than another if
+  # * the mount point of A is a leading substring of B's mount point
+  # * the mount point of A is a leading substring of B's device (B is a bind mount)
+  #
+  # @note Does not really care about directory names and does substring
+  #   comparison only. So /media will be put before /mediastuff. This is
+  #   not necessary but will be rare and usually cause no harm, either.
+  #
+  # @api private
+  def self.order(records)
```

Comment:
Great point about complexity, but at first glance, this looks to me as though it's basically a clever trick to stabilize Ruby's own sort. That's helpful, but we're still running the risk of critical comparisons never taking place.

Yes, this is a total order, but only because of the arbitrary "keep original order" ruling. It still does not address the issue that for the primary comparator, most elements will be equal.

In other words, if the quicksort happens to choose convenient pivot elements, this will work, but not in the average case. Unless I'm wrong :-) I'll pass it by the existing tests in any case.

Thanks for looking into this @peterhuene !